core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1972,62 +1972,62 @@ protected Type visitExtract(Extract node, StackableAstVisitorContext<Context> co
-                            !(type.equals(INTERVAL_DAY_TIME))) {
-                        throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
-                    }
-                    break;
-                case QUARTER:
-                case WEEK:
-                case DAY_OF_MONTH:
-                case DAY_OF_WEEK:
-                case DOW:
-                case DAY_OF_YEAR:
-                case DOY:
-                case YEAR_OF_WEEK:
-                case YOW:
-                    if (!(type instanceof DateType) &&
-                            !(type instanceof TimestampType) &&
-                            !(type instanceof TimestampWithTimeZoneType)) {
-                        throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
-                    }
-                    break;
-                case HOUR:
-                case MINUTE:
-                case SECOND:
-                    if (!(type instanceof TimestampType) &&
-                            !(type instanceof TimestampWithTimeZoneType) &&
-                            !(type instanceof TimeType) &&
-                            !(type instanceof TimeWithTimeZoneType) &&
-                            !(type.equals(INTERVAL_DAY_TIME))) {
-                        throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
-                    }
-                    break;
-                case TIMEZONE_MINUTE:
-                case TIMEZONE_HOUR:
-                    if (!(type instanceof TimestampWithTimeZoneType) && !(type instanceof TimeWithTimeZoneType)) {
-                        throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
-                    }
-                    break;
-                default:
-                    throw new UnsupportedOperationException("Unknown field: " + field);
+            if (Extract.Field.YEAR.equals(field)) {
+                checkExtractMonth(node, type, field);
+                outputType = INTEGER;
+            }
+            else if (Extract.Field.MONTH.equals(field)) {
+                checkExtractMonth(node, type, field);
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.DAY.equals(field)) {
+                if (!(type instanceof DateType) &&
+                        !(type instanceof TimestampType) &&
+                        !(type instanceof TimestampWithTimeZoneType) &&
+                        !(type.equals(INTERVAL_DAY_TIME))) {
+                    throw semanticException(TYPE_MISMATCH, node.getExpression(), "Cannot extract %s from %s", field, type);
+                }
+                outputType = TINYINT;
+            }
+            else if (Extract.Field.QUARTER.equals(field)) {
+                checkExtractYearOfWeek(node, type, field);
```

Comment:
This (and all other usages) are misleading. The field is `QUARTER`, yet the method says "checkExtractYearOfWeek`

Reply:
I agree. I've updated the method names to more meaningful names. Please do advise if there would be better names more fit in the corresponding context.